jb/itch5/array_based_order_book.hpp

```diff
@@ -0,0 +1,660 @@
+ * template classes order_book and compute_book:
+ * - usage: jb::itch5::order_book<jb::itch5::array_based_order_book>
+ */
+struct array_based_order_book {
+  using buys_t = array_based_book_side<std::greater<price4_t>>;
+  using sells_t = array_based_book_side<std::less<price4_t>>;
+  class config;
+};
+
+/**
+ * Configure an array_based_order_book config object
+ */
+class array_based_order_book::config : public jb::config_object {
+public:
+  config()
+      : max_size(
+            desc("max-size")
+                .help(
+                    "Configure the max size of a array based order book."
+                    " Only used when enable-array-based is set"),
+            this, jb::itch5::defaults::max_size) {
+  }
+
+  config_object_constructors(config);
+
+  /// Validate the configuration
+  void validate() const override {
+    if ((max_size() <= 0) or (max_size() > jb::itch5::defaults::max_size)) {
+      std::ostringstream os;
+      os << "max-size must be > 0 and <=" << jb::itch5::defaults::max_size
+         << ", value=" << max_size();
+      throw jb::usage(os.str(), 1);
+    }
+  }
+
+  jb::config_attribute<config, std::size_t> max_size;
+};
+
+/// price level limit between mill and penny
+int constexpr TK_DOLLAR = 10000;
+
+/**
+ * Represent one side of the book.
+ *
+ * This implementation uses a top_levels_ fix sized vector<price4_t> to
+ * keep the prices around the inside. We are hoping this makes a faster
+ * order book side (instead of map based).
+ *
+ * The top_levels_ range is kept by px_begin_top_ (worst price on the range)
+ * and px_end_top_ (best price on the range).
+ * If a price is worse than the px_begin_top_ is kept on the bottom_levels_
+ * (regular map).
+ * When a new inside crosses the limits of top_levels_, then the limits are
+ * redefined and prices are moved to or from bottom_levels_ depending on the
+ * direction the new inside is moving.
+ *
+ * The main motivation of this implementation is to have a second order book
+ * side type (in addition to map_based_order_book) in order for them to be
+ * tested/compare with the jaybeams FFT infrastructure.
+ *
+ * Class implementation of struct
+ * array_based_order_book buy and side types.
+ * @tparam compare_t function object class type to sort the side
+ *
+ */
+template <typename compare_t>
+class array_based_book_side {
+public:
+  class config;
+  explicit array_based_book_side(array_based_order_book::config const& cfg)
+      : better_()
+      , max_size_(cfg.max_size())
+      , top_levels_(cfg.max_size(), 0)
+      , bottom_levels_()
+      , px_inside_(empty_quote().first.as_integer())
+      , px_begin_top_(empty_quote().first.as_integer())
+      , px_end_top_(empty_quote().first.as_integer()) {
+  }
+
+  /// @returns an empty bid or offer based on compare function
+  /// empty bid for less, empty offer for greater.
+  half_quote empty_quote() const {
+    if (better_(price4_t(1), price4_t(0))) {
```

Comment:
I think you are right this time.

Excellent solution BTW. 

Working on it...